### PR TITLE
[MINI-4672] Download Files to device

### PR DESCRIPTION
**Pull request by laubengaier**
# Description
New implementation for downloading files to device passing a url and filename instead of xhr (and base64)

## Links
https://jira.rakuten-it.com/jira/browse/MINI-4672

# Checklist
- [x] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [ ] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [ ] I ran `fastlane ci` without errors

### CODE REVIEW

**Review by corycaywood — APPROVED** [review]
👍

**Review by rleojoseph — APPROVED** [review]
Looks good to me :+1:

**Review by Climbatize — APPROVED** [review]
LGTM 👍

**Review by Climbatize — COMMENTED** [review]
Please document and add a changelog line before merging

**Review by Climbatize — CHANGES_REQUESTED** [review]
I noticed that unit tests failed because of your changes, please ensure that your changes are retro-compatible with 4.0.0, put deprecations if needed

File: Sources/Classes/core/Protocols/MiniAppMessageDelegate.swift [hunk 1]
```diff
@@ -89,3 +96,7 @@ public class MAHostEnvironmentInfo: Codable {
         )
     }
 }
+
+public class DownloadHeaders: Codable {
+    let token: String?
+}
```
Comment (corycaywood) [inline Sources/Classes/core/Protocols/MiniAppMessageDelegate.swift:102]: @laubengaier Sorry, I missed seeing this before. The headers needs to be open ended and allow any number of generic String pair of headers, not just a single param.

File: Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift [hunk 2]
```diff
@@ -519,6 +521,24 @@ internal class MiniAppScriptMessageHandler: NSObject, WKScriptMessageHandler {
...
+            let fileName = parameters?.filename,
+            let url = parameters?.url,
+            let headers = parameters?.headers
+        {
+             hostAppMessageDelegate?.downloadFile(fileName: fileName, url: url, headers: headers) { (result) in
+                switch result {
+                case .success:
+                    self.executeJavaScriptCallback(responseStatus: .onSuccess, messageId: callbackId, response: fileName)
+                case .failure:
+                    self.executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: getMiniAppErrorMessage(MiniAppJavaScriptError.unexpectedMessageFormat))
```
Comment (corycaywood) [inline Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift:536]: I think we need to use a specific format for the Error, similar to what we did with `getAccessToken`. Also in the case that the download failed, we should pass the status code and message. This is what I suggested, but you can sync up with Android side:
```
{
  status: number,
  reason: String
}
```

**Review by corycaywood — CHANGES_REQUESTED** [review]

File: Sources/Classes/core/Protocols/MiniAppMessageDelegate.swift (hunk 1 above)
Comment (laubengaier) [reply Sources/Classes/core/Protocols/MiniAppMessageDelegate.swift]: ok gotcha, replaced it with [String: String]

File: Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift (hunk 2 above)
Comment (laubengaier) [reply Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift]: added a new error `MASDKDownloadFileError` that should cover the error cases

**Comment by laubengaier** [comment]
> I noticed that unit tests failed because of your changes, please ensure that your changes are retro-compatible with 4.0.0, put deprecations if needed

what do you specifically mean 😄 ? it's an addition before we could just intercept the tapped url

**Review by Climbatize — APPROVED** [review]
LGTM

**Review by corycaywood — APPROVED** [review]
Looks good 👍

**Pull request merged**